dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/DefaultTrackerObjectsDeletionService.java

```diff
@@ -0,0 +1,444 @@
+        this.programStageInstanceService = stageInstanceService;
+        this.relationshipService = relationshipService;
+        this.trackerAccessManager = trackerAccessManager;
+        this.enrollmentTrackerConverterService = enrollmentTrackerConverterService;
+        this.eventTrackerConverterService = eventTrackerConverterService;
+        this.notifier = notifier;
+    }
+
+    @Override
+    public TrackerTypeReport deleteEnrollments( TrackerBundle bundle, TrackerType trackerType )
+    {
+        TrackerTypeReport typeReport = new TrackerTypeReport( trackerType );
+
+        List<Enrollment> enrollments = bundle.getEnrollments();
+
+        for ( int idx = 0; idx < enrollments.size(); idx++ )
+        {
+            String uid = enrollments.get( idx ).getEnrollment();
+
+            boolean existsEnrollment = programInstanceService.programInstanceExists( uid );
+
+            TrackerObjectReport trackerObjectReport = new TrackerObjectReport( TrackerType.ENROLLMENT );
+
+            if ( existsEnrollment )
+            {
+                ProgramInstance programInstance = programInstanceService.getProgramInstance( uid );
+
+                if ( bundle.getUser() != null )
+                {
+                    List<TrackerErrorReport> trackerErrorReports = isAllowedToDeleteEnrollment( idx, bundle.getUser(), programInstance, bundle );
+
+                    if ( !trackerErrorReports.isEmpty() )
+                    {
+                        trackerObjectReport.getErrorReports().addAll( trackerErrorReports );
+                        trackerObjectReport.setIndex( idx );
+                        trackerObjectReport.setUid( uid );
+
+                        typeReport.addObjectReport( trackerObjectReport );
+                        typeReport.getStats().incIgnored();
+                        return typeReport;
+                    }
+                }
+
+                List<Event> events = eventTrackerConverterService.to( Lists.newArrayList( programInstance.getProgramStageInstances() ) );
+
+                TrackerBundle trackerBundle = TrackerBundle.builder().events( events ).user( bundle.getUser() ).build();
+
+                // Associated events should be deleted provided user has authority for that.
+                TrackerTypeReport eventReport = deleteEvents( trackerBundle, TrackerType.EVENT );
+
+                if ( !eventReport.isEmpty() )
+                {
+                    trackerObjectReport.getErrorReports().addAll( eventReport.getErrorReports() );
+                }
+
+                programInstanceService.deleteProgramInstance( programInstance );
+                teiService.updateTrackedEntityInstance( programInstance.getEntityInstance() );
+
+                typeReport.addObjectReport( trackerObjectReport );
+                typeReport.getStats().incDeleted();
+            }
+            else
+            {
+                trackerObjectReport.getErrorReports().add( TrackerErrorReport.builder()
```

Comment:
It wouldnt be easier to understand the flow if we just validate also the delete payload and make assumptions when actually deleting?
It seems to me that all this errors that we are raising here would be covered by the validation.
Are there some false errors that would be raised?

Reply:
fixed